### ISSUE TITLE
AnimatedSprite.playing and .frame returns Nil.

### ISSUE THREAD

**Issue by Ybaedbaenn** [godotengine/godot#50373]
### Godot version

3.3.2.stable.oficial

### System information

Windows 10, GLES3, NVIDIA GeForce GT 540M, graphics driver version 23.21.13.9135, Intel(R) Core(TM) i7-2670QM CPU @ 2.20GHz

### Issue description

Accessing AnimatedSprite.playing and AnimatedSprite.playing can return Nil. In a certain situation changing AnimatedSprite animation suspends the program.
![issue](https://user-images.githubusercontent.com/84321760/125203532-02279200-e279-11eb-896f-aaf9cbbf32e7.png)

### Steps to reproduce

I have included a minimal reproduction project. When running project, it breaks, and shows error: cannot convert Nil to String, at line: push_warning($AnimatedSprite.playing).
After deleting the line, project breaks, and shows the same error at line: push_warning($AnimatedSprite.frame)
After deleting both lines, project suspends while executing.

### Minimal reproduction project

[AnimatedSprite_returns_Nil.zip](https://github.com/godotengine/godot/files/6797296/AnimatedSprite_returns_Nil.zip)

**Comment by Ybaedbaenn** [comment]
Possible connection with bug described hehre: https://www.reddit.com/r/godot/comments/gzk19l/end_of_animation_function_called_twice/

**Comment by Ybaedbaenn** [comment]
Possible conection with the bug described here: https://www.reddit.com/r/godot/comments/gzk19l/end_of_animation_function_called_twice/

**Comment by KoBeWi** [comment]
It doesn't hang when you change the last line to `$AnimatedSprite.set_deferred("animation", "a")`. Seems like the animation finished signal is emitted immediately when switching animation, because it has 0 length, which causes the program to infinite-loop or something.

**Comment by Ybaedbaenn** [comment]
I tried to use print(), but it printed nothing.

**Comment by Ybaedbaenn** [comment]
I tried to use print(), but it printed nothing.
![picture](https://user-images.githubusercontent.com/84321760/125208573-9acb0b80-e293-11eb-93fa-b1435e61d884.png)

**Comment by KoBeWi** [comment]
It prints when you use the `call_deferred()` workaround I mentioned.

**Comment by Ybaedbaenn** [comment]
That's right. However, regarding the infinite loop, by adding line: 
if $AnimatedSprite.animation != "a": 
   $AnimatedSprite.animation = "a"
I made sure that the animation would only be switched to "a" once(not infinitely), but if I don't use set_deferred(), the program still hangs.

**Comment by Ybaedbaenn** [comment]
That's right. However, regarding the infinite loop, by adding line:
if $AnimatedSprite.animation != "a":
   $AnimatedSprite.animation = "a"
I made sure that the animation would only be switched to "a" once(not infinitely), but if I don't use set_deferred(), the program still hangs.
![picture](https://user-images.githubusercontent.com/84321760/125208903-a28baf80-e295-11eb-85cf-75f3c6617914.png)

**Comment by KoBeWi** [comment]
I looked at the source and the problem is changing animation during finished signal (it modifies some variables that are used after emitting the signal and it causes invalid state of AnimatedSprite). We either should disallow doing it or make sure that it doesn't hang.

**Comment by Ybaedbaenn** [comment]
It seems to me that there is also a problem with print(), because as the first part of the function, it should have printed before the program suspended, which did not happen (unlike push_warning() in the console)

**Comment by KoBeWi** [comment]
Yeah if the program hangs, the print doesn't flush. You can still see the values in the external console.

**Comment by Ybaedbaenn** [comment]
Is this normal or a known issue?

**Comment by Ybaedbaenn** [comment]
Is this normal or a known issue? Or does print just execute at the end of the function?

**Comment by KoBeWi** [comment]
I don't know the internals of `print()`, but it stops working when you cause infinite loop in C++ code. It rarely happens, because we should have safeguards against this (e.g. not being able to call a method with invalid arguments etc).

**Comment by Ybaedbaenn** [comment]
This is still a problem for debugging, if you want the function that crashes the program to output something before it does.

**Comment by Ybaedbaenn** [comment]
> 
> 
> I looked at the source and the problem is changing animation during finished signal (it modifies some variables that are used after emitting the signal and it causes invalid state of AnimatedSprite). We either should disallow doing it or make sure that it doesn't hang.
> 
> EDIT:
> Although this is probably unrelated to the null issue. It doesn't make sense that you can print the value, but push_warning fails.

If it is about not being able to set FPS to zero, I will explain, that (in my other project) I used this technique to take advantage of the second animation as a texture when some object is not clicked by a user for a long time. It is easier than adding separate Nodes for Sprite and AnimatedSprite, and setting FPS to zero is to avoid updating the object unnecessarily.

**Comment by Ybaedbaenn** [comment]
> 
> 
> I looked at the source and the problem is changing animation during finished signal (it modifies some variables that are used after emitting the signal and it causes invalid state of AnimatedSprite). We either should disallow doing it or make sure that it doesn't hang.
> 
> EDIT:
> Although this is probably unrelated to the null issue. It doesn't make sense that you can print the value, but push_warning fails.

If it is about not being able to set FPS to zero, I will explain, that (in my other project) I used this technique, to take advantage of the second animation as a texture; when some object is not clicked by a user for a long time. It is easier than adding separate Nodes for Sprite and AnimatedSprite, and setting FPS to zero is to avoid updating the object animation unnecessarily.

**Comment by timothyqiu** [comment]
> The null part is weird though. When you use `print()` instead, it correctly prints `true`.

`push_warning` expects its argument to be a String. It works fine if written as `push_warning(str($AnimatedSprite.playing))`.

Maybe it's the error message that reports the wrong type.

**Comment by Ybaedbaenn** [comment]
The bug still exists in Godot version 3.3.4.stable.official.

**Comment by Ybaedbaenn** [comment]
I have made a new, analogous **4.1** reproduction project, and it seems that the issue is gone.
[AnimatedSpriteNullTest4.1.zip](https://github.com/godotengine/godot/files/11967248/AnimatedSpriteNullTest4.1.zip)

The only change that may be needed is adding information about setting fps to 0 or using set_animation_speed("name", 0) in the [SpriteFrames'](https://docs.godotengine.org/en/stable/classes/class_spriteframes.html#class-spriteframes-method-set-animation-speed) documentation, something like the one [AnimatedSprite2D's](https://docs.godotengine.org/en/stable/classes/class_animatedsprite2d.html#class-animatedsprite2d-property-speed-scale) speed_scale property already has.

###### Version: v4.1.stable.official.970459615
###### Graphics: OpenGL API 3.3.0 NVIDIA 391.35 - Compatibility - Using Device: NVIDIA - GeForce GT 540M
